Submodule browse: Select both first/second commits

gerhardol opened the pull request:
Related to discussions in #8700 how to handle diffs for submodules

## Proposed changes

GE has a decent overview of changes in a submodule, but sometimes I want to see the actual changes in files or more than first/selected commit message. A git-diff --dir-diff will not work with all difftools, not show working dir (unstaged) changes and no intermediate commit messages.

The best difftool is GE itself, when a subrepo is currently opened only the selected commit is selected though. It can be difficult to find the first submodule commit. In the screenshot it was at least the same page, harder if there are more changes.

## Screenshots <!-- Remove this section if PR does not change UI -->

Open submodule in e9eff4bb646ccd355f221c81cbf98e1eeab5a3b4 (or doubleclick).
If the WorkTree was browsed (Commit, RevDiff, side panel later), OldCommit->Worktree would be displayed instead after the change.

![image](https://user-images.githubusercontent.com/6248932/103275158-d111ff00-49c3-11eb-912e-71550fff3e5b.png)

Before | After
-------|---------
![image](https://user-images.githubusercontent.com/6248932/103275284-24844d00-49c4-11eb-9f3d-f98805f40356.png) | ![image](https://user-images.githubusercontent.com/6248932/103275243-0fa7b980-49c4-11eb-93db-7c8afa1e3edc.png)

## Test methodology <!-- How did you ensure quality? -->

Test updated.

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -795,13 +795,16 @@ private async Task OpenSubmoduleAsync()
                 {
                     FileName = Application.ExecutablePath,
-                    Arguments = "browse -commit=" + status.Commit,
+                    Arguments = $"browse -commit={selected},{status.OldCommit}",
```
This code is identical to RevisionDiffControl.cs L459-474. Should we factor out a new method?

gerhardol replied:
I considered that, but settled with aligning the code
Where to put it?

```
        public async Task OpenSubmoduleAsync(FileStatusItem item, IFullPathResolver fullPathResolver)
        {
            var submoduleName = item.Item.Name;

            var status = await item.Item.GetSubmoduleStatusAsync().ConfigureAwait(false);
            var selected = item.SecondRevision.ObjectId == ObjectId.WorkTreeId
                ? item.SecondRevision.ObjectId
                : status.Commit;

            var process = new Process
            {
                StartInfo =
                {
                    FileName = Application.ExecutablePath,
                    Arguments = $"browse -commit={selected},{status.OldCommit}",
                    WorkingDirectory = _fullPathResolver.Resolve(submoduleName.EnsureTrailingPathSeparator())
                }
            };

            process.Start();
        }
```

RussKie replied:
❗  `IFullPathResolver fullPathResolver` must not be passed as an input argument

mstv commented on GitUI/GitUICommands.cs:
```diff
@@ -1609,13 +1609,35 @@ private bool RunBrowseCommand(IReadOnlyList<string> args)
             Console.Error.WriteLine($"No commit found matching: {arg}");
             return false;
+
+            static bool tryGetObjectIds(string arg, GitModule module, out List<ObjectId> objectIds)
```
?
```suggestion
            static bool TryGetObjectIds(string arg, GitModule module, out List<ObjectId> objectIds)
```

gerhardol replied:
OK

mstv commented on GitUI/CommandsDialogs/FormCommandlineHelp.resx:
```diff
@@ -118,7 +118,7 @@
   </resheader>
   <data name="_NO_TRANSLATE_commands.Text" xml:space="preserve">
-    <value>browse [path] [-filter=] [-commit=]
+    <value>browse [path] [-filter=] [-commit=&lt;sha1&gt;[,&lt;sha2&gt;]]
```
Unfortunately, the argument option `-filter=` from <= 2012 is inconsistent with most of the others.
So rather?
```suggestion
    <value>browse [path] [-filter=] [--commit &lt;sha1&gt;[,&lt;sha2&gt;]]
```

gerhardol replied:
-commit is also an old option, cannot be changed (probably used from shell extension too)

RussKie replied:
I think `--commit=abc123,09982039cbs00808dsfsd08978s` is totally fine

gerhardol replied:
If --commit (or --selected-commit and --first-commit) is added then -commit should be handled still. Maybe shell extension could be synchronized (not for me that only run private builds...) but vsix and other users could be affected, I do not see all risks. It is hard to change published interfaces.

RussKie replied:
This `[-commit=&lt;sha1&gt;[,&lt;sha2&gt;]]` looks like an additive change - i.e. the original functionality should be unaffected. We just allow to append another hash to it.
Am I overlooking something else?

gerhardol replied:
Modifying arguments for -commit is compatible.
@mstv remarked on the other arguments using GNU/Posix standard that single char options have one dash and can often be combined (ls -ar is the same as ls -a -r) and multicharacter options has two dashes (that cannot be combined) (ls --all --reverse).

mstv replied:
> @mstv remarked on the other arguments using GNU/Posix standard that ... and multicharacter options has two dashes.

But I missed that `-commit=...` has already existed.
Since GE is mainly a GUI tool, I think a cleanup of the sligthly inconsistent format of the CLI arguments is not worth to be cleaned up.

RussKie commented on GitUI/GitUICommands.cs:
```diff
@@ -1609,13 +1609,35 @@ private bool RunBrowseCommand(IReadOnlyList<string> args)
             Console.Error.WriteLine($"No commit found matching: {arg}");
             return false;
+
+            static bool tryGetObjectIds(string arg, GitModule module, out List<ObjectId> objectIds)
```
We unnecessary allocate memory here - we expect maximum of two items, yet we allocate a dynamic list, and then re-allocate the list into an array.
Allocate and return `ObjectId[2]`

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -162,7 +162,7 @@ private RefFilterOptions RefFilterOptions
 
         [CanBeNull]
-        internal ObjectId InitialObjectId { private get; set; }
+        internal ObjectId[] InitialObjectIds { private get; set; }
```
Passing arrays around is inherently dangerous, do we expect more than 2 commits to be selected? If not, should we have two properties instead?

gerhardol replied:
More than two is an edge case, like if you want to see a submodule selected with the same commits you have multiselected or to select multiple commits from the cmd line. But that would be separate extensions.

Would passing a list around be OK?

Otherwise new arguments --selected-commit and --first-commit should be added (with backwards compatibility with -commit) should be added

RussKie replied:
If automatically selecting more than two commits is an edge case I propose we have two `ObjectId` properties, thus removing the need to pass arrays around.
I propose we keep `InitialObjectId` as is (at least its functionality, but happy to have it renamed to something like `FirstSelectedCommit`), and add another property  for the second selected commit.

gerhardol replied:
I changed this even if I find that this makes the code harder to maintain.

RussKie commented:
@drewnoakes was there any reason not to make `ObjectId` a `struct`? After all it looks like 5 uints, that can leave on the stack.

RussKie reviewed (changes requested):
My only concern is that we are extremely inconsistent in naming: `InitialObjectId`, `FirstObjectId`, `selectedId`, `firstId`, `selectedCommit`, `firstCommit`, `selectedSha1`, `firstSha2`...

We need to be consistent, otherwise this will turn into a undecipherable goo very quickly.
Xml-docs annotating input args would help as well.

gerhardol commented:
squashed and rebased on master

* Updated review comments
* Changed handling of displaying new submodules (would show diff from worktree to the current commit which may not be expected)
* Removed handling to use empty commits as worktree (use the artificial sha if needed)

RussKie commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -122,7 +122,14 @@ private FormBrowse()
+        /// <param name="commands">commands in the current form</param>
+        /// <param name="filter">filter to apply to browse</param>
+        /// <param name="selectedId">Currently selected commit id</param>
+        /// <param name="firstId">First commit id in a diff</param>
```
Should we call it `diffId` instead?

gerhardol replied:
(selected|first)(Id|Commit) is the common name, except in fileStatusList where First/Second is used (as Selected is not always Second). (Id for ObjectId and Commit for GitRevision).

"selected" is the primary, last selected commit which normally is the second in a diff (to or B)
"first" is the first in a diff, there can be several first before the selected though.

RussKie commented:
FirstSelectedCommitId and LastSelectedCommitId?

A bit verbose perhaps but provide a good hint to developers.

On Mon, 4 Jan. 2021, 11:09 am Gerhard Olsson, <notifications@github.com>
wrote:

> *@gerhardol* commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/FormBrowse.cs
> <https://github.com/gitextensions/gitextensions/pull/8703#discussion_r551072968>
> :
>
> > @@ -122,7 +122,14 @@ private FormBrowse()
>              InitializeComplete();
>          }
>
> -        public FormBrowse([NotNull] GitUICommands commands, string filter, ObjectId selectCommit = null)
> +        /// <summary>
> +        /// Open Browse - main GUI including dashboard
> +        /// </summary>
> +        /// <param name="commands">commands in the current form</param>
> +        /// <param name="filter">filter to apply to browse</param>
> +        /// <param name="selectedId">Currently selected commit id</param>
> +        /// <param name="firstId">First commit id in a diff</param>
>
> (selected|first)(Id|Commit) is the common name, except in fileStatusList
> where First/Second is used (as Selected is not always Second). (Id for
> ObjectId and Commit for GitRevision).
>
> "selected" is the primary, last selected commit which normally is the
> second in a diff (to or B)
> "first" is the first in a diff, there can be several first before the
> selected though.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8703#discussion_r551072968>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVASL3IMXKANUAFTI3SYEBLLANCNFSM4VNBO6IA>
> .
>

drewnoakes commented:
> @drewnoakes was there any reason not to make ObjectId a struct? After all it looks like 5 uints, that can leave on the stack.

Only reason I can think of is the potential for excessive copying. I don't think that'd be an issue in practice, and can be amerliorated by newer C# features such as `readonly struct` and `in` parameter modifiers.

Conversion from class to struct would need to take nulls into account.

drewnoakes commented:
If someone makes this conversion, please tag me on the PR so I can take a look 👍

gerhardol commented:
> FirstSelectedCommitId and LastSelectedCommitId? A bit verbose perhaps but provide a good hint to developers.

FirstSelectedId and LastSelectedId in that case as Commit is normallt used for GitRevision.
This will deviate from the normal use use of selectedId/firstId, maybe OK as this is primarily for selection

gerhardol commented:
Added the handling to side panel too.
If someone agrees, I will add that to this PR too, otherwise it will be a new PR (datastructures changes slightly for submodules in DetailedSubmoduleInfo). Of course, naming depends on the discussion here.

https://github.com/gerhardol/gitextensions/pull/new/feature/browse-submodules-2

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -162,7 +162,8 @@ private RefFilterOptions RefFilterOptions
 
         [CanBeNull]
-        internal ObjectId InitialObjectId { private get; set; }
+        internal ObjectId SelectedId { private get; set; }
+        internal ObjectId FirstId { private get; set; }
```
Please add xml-docs here as well

gerhardol replied:
Done, squashed

RussKie commented:
> Added the handling to side panel too.
> If someone agrees, I will add that to this PR too, otherwise it will be a new PR (datastructures changes slightly for submodules in DetailedSubmoduleInfo). Of course, naming depends on the discussion here.
> 
> [gerhardol/gitextensions/pull/new/feature/browse-submodules-2](https://github.com/gerhardol/gitextensions/pull/new/feature/browse-submodules-2)

I think it'd be easier if we merge this, and then look at the next one. Otherwise we may bog down.

gerhardol commented:
> I think it'd be easier if we merge this, and then look at the next one. Otherwise we may bog down.

I agree. In some other changes the follow up PR could change the original PR, but this is even separate files.

gerhardol commented:
@msftbot merge in 48 hours
